File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/constant/parts/BinaryPart.java
package xyz.wagyourtail.commons.parsers.impl.constant.parts;

import lombok.val;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

public class BinaryPart extends StringData.OnlyRaw<Data.SingleContent<String>> {

    private BinaryPart(String content) {
        super(content, BinaryPart::getContentChecked);
    }

    public BinaryPart(CharReader<?> reader) {
        super(reader, BinaryPart::getContentChecked);
    }

    public static BinaryPart parse(String rawContent) {
        StringCharReader reader = new StringCharReader(rawContent);
        val number = new BinaryPart(reader);
        reader.expectEOS();
        return number;
    }

    public static BinaryPart unchecked(String rawContent) {
        return new BinaryPart(rawContent);
    }

    private static SingleContent<String> getContentChecked(CharReader<?> reader) {
        val first = reader.peek();
        if (first != '0' && first != '1') {
            throw reader.createException("Not a binary number: " + (char) first);
        }

        return new SingleContent<>(reader.takeWhile(e -> e == '0' || e == '1'));
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/constant/number/BinaryPart.kt
package xyz.wagyourtail.commonskt.parsers.impl.constant.number

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader

class BinaryPart(rawContent: String) : StringData.OnlyRaw<Data.SingleContent<String>>(rawContent) {

    constructor(content: SingleContent<String>) : this(content.toString())

    companion object : StringDataBuilder<BinaryPart, SingleContent<String>> {

        override fun invoke(rawContent: CharReader<*>): BinaryPart {
            return BinaryPart(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
            val first = reader.peek()
            if (first != '0' && first != '1') {
                throw reader.createException("Not a binary number: $first")
            }
            return SingleContent(reader.takeWhile { it == '0' || it == '1' })
        }

    }

    override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
        return Companion.checkedBuildContent(reader)
    }

}